### DISCUSSION TITLE
Migration from simple json data source plugin

### DISCUSSION THREAD

**Discussion by yesoreyeram** [grafana/grafana-infinity-datasource#740] [Docs]
<p align="center">
<img src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/7319d1c5-8e5a-4a4a-bf2f-66e37f9a98f5" width="100" height="100" /><b>=></b>
<img src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/9f38194c-0033-4764-8553-0c5f0268ab4d"  width="120" height="120" />
</p>

[Grafana's simple JSON datasource plugin](https://grafana.com/grafana/plugins/grafana-simple-json-datasource/) is a legacy way of connecting JSON endpoints from Grafana. This requires you to run your own http server that implements certain HTTP endpoints such as `/`, `/search`, `/query` and `/annotations` endpoint and require to return data in certain format. This plugin was build for older versions of Grafana and no longer maintained. Also this plugin was built using legacy AngularJS technology which is now deprecated in recent versions of Grafana. This makes this plugin unusable anymore.

## Migrating to infinity plugin

> [!IMPORTANT]
> Infinity is not a drop-in replacement for simple JSON datasource plugin. This require manual migration efforts. But migrating to infinity is strongly recommended approach if you are using simple JSON API datasource plugin.

### Approach 1 : Direct API connection / Simple approach / Recommended approach

This is much **easier and recommended** approach to connect JSON api endpoints directly. Infinity allows you to connect directly to your JSON endpoints instead of requiring you to write your server implementation comparing to grafana simple json server approach. Refer [Infinity plugin website](https://grafana.com/grafana/plugins/yesoreyeram-infinity-datasource/) for more details about connecting your APIs directly.

With this approach, you can get rid of your custom json server and directly connect your API endpoints via Infinity plugin. If this approach is not possible for any reason, use the below alternate migration approach.

### Approach 2 : Migrating using Grafana simple json server approach

With this approach, Instead of connecting your API endpoints directly, you will be connecting via grafana simple JSON server (legacy server). Also in grafana, instead of using simple json datasource plugin, you will be using Infinity plugin. 

#### Migrating the configuration

In terms of configuration editor, there is no much change. The URL used in simple json datasource goes into infinity config misc/url section. Other options such as proxy, tls/ca certificates, headers goes into network and headers section.

| Before | After |
|--|--|
| <img width="817" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/f7381eaa-8791-49e1-b114-d69e40876f26"> | <img width="881" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/f064f74c-7302-4518-8102-4e0a0c92a500">|

#### Provisioning

Provisioning of simple json and infinity is almost similar only the plugin id differs. Refer the infinity plugin [provisioning documentation](https://grafana.com/docs/plugins/yesoreyeram-infinity-datasource/latest/setup/provisioning/) for more examples.

| Before | After |
|--|--|
| <img width="424" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/455cd3e0-2e71-4f26-8a66-6da978532609"> | <img width="417" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/ff63244e-5ba2-4642-9321-8efddd8898a0">|

#### Migrating the queries ( quick migration / frontend parser )

If you are using one or more queries using Simple JSON as shown below, you can migrate to infinity using the steps provided below. This approach is simple but less powerful. Doesn't support grafana backend features such as alerting.

| Before | After |
|---|---|
| <img width="1326" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/2d26e1cc-2ed1-488e-9607-0bd5226785a9">| <img width="1335" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/be3e07a9-4d7b-45b0-9cb6-e7ef86b74d99"><img width="1299" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/74cc2435-4bfb-444d-83ce-62198a46e0bc">|
| Select the target in the query editor. Example `upper_25`| Select `JSON` as query type. <br/>Select `Default`/`Frontend` as your parser. <br/>Source: `URL`. <br/>Format: `As IS`/`Legacy`. <br/>HTTP Method: `POST`. <br/>URL : `/query`<br/>HTTP Body: In your http body you need to specify the targets: `{  "targets": [{ "target":"upper_25" }] }`| 

#### Migrating the queries ( recommended migration / backend parser )

Use this approach to migrate to get support for features such as alerting, public dashboards, query caching etc.

| Before | After |
|---|--|
|<img width="1319" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/2a619d89-d15f-4bc0-a6c3-d958ec2d74da">|<img width="1314" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/04e531c3-00f9-44de-8429-0850a4541cad"><img width="876" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/f1771ff6-fa93-4c43-8933-5727e2116c31">|
|Select the target in the query editor. Example `upper_25`|Query Type: `URL` <br/>Parser: `Backend` <br/>Source: `URL` <br/>Format: `Time series` <br/>HTTP method: `POST` <br/>URL: `/query` <br/>BODY: `{  "targets": [{ "target":"upper_25" }]}` <br/>Parsing options/Root: `datapoints` <br/> Column 1: `0` as selector. `upper_25` as alias. `Number` as format.<br/>Column 2: `1` as selector. `Time` as alias. `Unix (ms)` as format.<br/>|

### Migrating the annotations

Migrating the annotations is similar to query. For example, below screenoshots show different annotation creations

###### Using Simple JSON
<img width="631" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/471fffe5-ef35-43b0-93a1-c6e666244ef4">

###### Using Infinity
<img width="1385" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/216138f6-b7d7-4911-93fb-f30da94b9efe">

#### Troubleshooting

> [!NOTE]
> Infinity doesn't support direct browser connection to your API endpoints. All the requests will be proxied from grafana server.

#### Sample migration dashboard

<img width="1792" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/24038bad-2dbf-474a-b0e6-63405092057a">

**Comment by peret2000** [comment]
Thanks a lot for this guide for migration. Unfortunately, images cannot be opened or saved to file, so it is really difficult to have a look at them. If you could give access to the images, it would be great. Thanks!

**Reply by yesoreyeram** [reply]
Have you tried right click and save the image? Or just use your browser zoom. ( Am I missing something else )...?

**Reply by peret2000** [reply]
I tried both methods (and also copying the link). But for some reason, just afeter writing the comment I could see the images. I thought you arranged it very fast, or thta the fact of writing the comment did the trick (!!!). But probably, neither of them was the issue, simply there was some mitake in my browser. But anyway, it is ok. Sorry for the confusion

**Reply by james00002** [reply]
I have the same problem with the images. Some of them are much to tiny to read, even if I use maximum browser zoom.
If I open the images in a new tab, I usually get a 404 error from githubusercontent.com but occasionally, seemingly at random, they will open up okay.

**Reply by peret2000** [reply]
Some time later, I found the error is due to the default or active google user in the browser (using Chrome). It tried to open github with a user that perhaps is not authenticated in github. Switching the user, or refreshing this page did the trick

**Comment by james00002** [comment]
There really needs to be a way to extract the "Columns" in the "Parsing options & Result fields" section from the same Json format that simple-json-datasource supported.

Without that feature, Infinity isn't really usable as a replacement.

Suppose a simple-json target had 20 columns defined. All of those showed up with grafana datatypes in the "columns" section of the response. Now it is necessary to manually type in those 20 column definitions for every single panel that uses the target :(

**Reply by weakiwi** [reply]
Hello James
  Do u have any solutions for this `dynamic columns` issue? Now I have a query like `sum latency{} by (instance)`, I have no idea how to deal with it.
  
 Thx!

**Reply by james00002** [reply]
No, I never found a solution. I just ended up manually typing 20 column definitions in multiple panels 😞

**Reply by weakiwi** [reply]
my solution is converting the line graph to table, cuz my column is dynamic(e.g: api/url_path) so I can't enum all the column.

**Comment by Lofesa** [comment]
Hi to all
I come here to see if anyone can guide or give some clue about migrating from grafana-simple-json-datasource.
In previuos versions of grafana I get the plugin grafana-simple-json-datasource working whitout issue, but in grafana 11 I get messages about Angular deprecation. I managed to get grafana-simple-json-datasource working but trying to migrate to Infinity.

The scenario is:

I have several plesk server, they have internal garafana to collect data. I used this doc to monitoring several plask server
https://www.plesk.com/kb/support/how-to-monitor-several-plesk-servers-with-grafana-extension-2/

As stated in the doc, I use an url like
https://hostname:8443/modules/monitoring/public/index.php/NyuIz_uQ8oZ6SKw31eSCxNt-KAWDlQ8e6VScvyqt
The last part of the url is an AUTH_TOKEN .

In the grafana-simple-json-datasource I put this url without any auth and then in the queries for panels I was able to see all the fields returned from the plesk server. But this doesn't work with Infinity. The error say is a 400 Bad request.
I tried to playing with the infinity plugin options but can't get it working.

Someone have any clue on how to configure this scenario?

Thanks in advance and sorry if that's no the place for posting this question.

**Reply by yesoreyeram** [reply]
Unfortunately it doesn't look like a valid/secure URL. I have never seen someone putting their token as part of the URL. :( 

What error you are getting with infinity. Have you pasted your entire URL `https://hostname:8443/modules/monitoring/public/index.php/NyuIz_uQ8oZ6SKw31eSCxNt-KAWDlQ8e6VScvyqt` ?

**Reply by Lofesa** [reply]
Yes, I've pasted the  entire url and get the 400 Bad request error in the Infinity but it works with the Simple Json.
In the Simple Json, using this url then I get the fields list in the queries so the >Simple Json is able to manage the response, but Infinity not.

**Comment by ArgArakelyan** [comment]
What version do I need for /query? /search works for me with Infinity, but /query gives me an internal server error 500.

The Infinity version is fairly old: 2.11.2.
